thrust/system/cuda/detail/unique.h

```diff
@@ -794,6 +804,36 @@ unique(execution_policy<Derived> &policy,
   return cuda_cub::unique(policy, first, last, equal_to<input_type>());
 }
 
+
+template <typename BinaryPred>
+struct zip_adj_not_predicate {
+  template <typename TupleType>
+  bool __host__ __device__ operator()(TupleType&& tuple) {
+      return !binary_pred(thrust::get<0>(tuple), thrust::get<1>(tuple));
+  }
+  
+  BinaryPred binary_pred;
+};
+
+
+__thrust_exec_check_disable__
+template <class Derived,
+          class ForwardIt,
+          class BinaryPred>
+typename thrust::iterator_traits<ForwardIt>::difference_type __host__ __device__
+unique_count(execution_policy<Derived> &policy,
+       ForwardIt                  first,
+       ForwardIt                  last,
+       BinaryPred                 binary_pred)
+{
+  if (first == last) {
+    return 0;
+  }
+  auto size = thrust::distance(first, last);
+  auto it = thrust::make_zip_iterator(thrust::make_tuple(first, first + 1));
+  return 1 + thrust::count_if(policy, it, it + (size - 1), zip_adj_not_predicate<BinaryPred>{binary_pred});
```

Comment:
I'm not sure: Does zip_iterator inherit the forward_iterator category from `first`? If so, then we probably need something like
```suggestion
  return 1 + thrust::count_if(policy, it, thrust::advance(it, size - 1), zip_adj_not_predicate<BinaryPred>{binary_pred});
```

Reply:
Yes, please use `thrust::advance` instead. Note that it doesn't return a new iterator, but rather increments its argument, so the suggested fix will need to be refactored a bit.

Reply:
right, I mixed up std::next and std::advance. Should Thrust have the former? Probably easy to add, too.